server/util/flagutil/types/types.go

```diff
@@ -195,20 +221,29 @@ func (f *FlagAlias) AliasedName() string {
 	return f.name
 }
 
-func (f *FlagAlias) AliasedType() reflect.Type {
-	flg := common.DefaultFlagSet.Lookup(f.name)
-	t, err := common.GetTypeForFlag(flg)
-	if err != nil {
-		return reflect.TypeOf(flg.Value)
-	}
-	return t
+func (f *FlagAlias) WrappedValue() flag.Value {
+	return common.DefaultFlagSet.Lookup(f.name).Value
 }
 
-func (f *FlagAlias) YAMLTypeAlias() reflect.Type {
-	flg := common.DefaultFlagSet.Lookup(f.name)
-	t, err := flagyaml.GetYAMLTypeForFlag(flg)
-	if err != nil {
-		return reflect.TypeOf(flg.Value)
-	}
-	return t
+type DeprecatedFlag struct {
+	flag.Value
+	name string
+}
+
+func DeprecatedVar[T any](value flag.Value, name string, usage string) *T {
+	common.DefaultFlagSet.Var(&DeprecatedFlag{value, name}, name, "**DEPRECATED** "+usage)
+	return reflect.ValueOf(value).Convert(reflect.TypeOf((*T)(nil))).Interface().(*T)
+}
+
+func (d *DeprecatedFlag) Set(value string) error {
+	log.Warningf("Flag %s is deprecated and was set on the command line.", d.name)
+	return d.Value.Set(value)
+}
+
+func (d *DeprecatedFlag) WrappedValue() flag.Value {
+	return d.Value
+}
+
+func (d *DeprecatedFlag) SetValueHook() {
+	log.Warningf("Flag %s is deprecated and was set programmatically, likely through the YAML config.", d.name)
```

Comment:
nbd if it's too complicated w/ the current design, but is there a way to differentiate whether this was set via YAML vs. something else? Thinking the error message would be  more clear for end users if we could remove the "and was set programmatically" part.

Also, can we print the flag help string here so that users know what the migration path is? e.g.

`log.Warningf("Flag %q (set via YAML config) is deprecated: %s", name, migrationPath)`

e.g.

`Flag "foo" (set via YAML config) is deprecated: no longer has any effect, please remove`